File: src/lib/watermark/contact-sheet.ts
import { normalizeModel } from "@/lib/exif/brand";
import type { FrameParams, TemplateConfig } from "@/stores/types";
import {
  formatTakenAt,
  getPreviewFontFamily,
  sanitizeCustomLines,
} from "./classic-bottom";
import {
  dprFor,
  fitPhoto,
  getCanvasRatio,
  paramsLine,
  type ShowcasePhoto,
} from "./renderer-utils";

const SHEET_BG = "#0a0a0a";
const SHEET_INK = "#f5f5f5";
const SHEET_MUTED = "#a8a8a8";
const SHEET_FRAME = "#ffffff";

/**
 * Contact sheet / proof contact print. Black background with the photo
 * matted by a thin white "frame number" border. Top + bottom edges carry a
 * row of sprocket holes like 35mm film. Bottom-right shows the EXIF.
 */
export function drawContactSheetPreview(
  canvas: HTMLCanvasElement,
  image: HTMLImageElement,
  _logo: HTMLImageElement | null,
  photo: ShowcasePhoto,
  frameParams: FrameParams,
  config: TemplateConfig,
) {
  const ctx = canvas.getContext("2d");
  if (!ctx) return;

  const dpr = dprFor(canvas);
  const baseWidth = 900;
  const ratio = getCanvasRatio(frameParams.canvasRatio, frameParams.canvasOrientation);
  const canvasW = baseWidth;
  const canvasH = canvasW / ratio;
  const watermarkActive = config.showWatermark ?? true;

  const perfH = Math.max(18, canvasH * 0.045);
  const sideMargin = canvasW * 0.07;
  const captionH = watermarkActive ? Math.max(78, canvasH * 0.14) : 0;

  const photoArea = {
    x: sideMargin,
    y: perfH + canvasH * 0.025,
    w: canvasW - sideMargin * 2,
    h: canvasH - perfH * 2 - captionH - canvasH * 0.06,
  };
  const placed = fitPhoto(photoArea, photo.width, photo.height);

  canvas.width = Math.round(canvasW * dpr);
  canvas.height = Math.round(canvasH * dpr);
  canvas.style.aspectRatio = `${canvasW} / ${canvasH}`;
  ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
  ctx.clearRect(0, 0, canvasW, canvasH);

  ctx.fillStyle = SHEET_BG;
  ctx.fillRect(0, 0, canvasW, canvasH);

  drawSprocketRow(ctx, 0, 0, canvasW, perfH);
  drawSprocketRow(ctx, 0, canvasH - perfH, canvasW, perfH);

  // White photo-mat frame.
  const matPad = Math.max(3, canvasW * 0.005);
  ctx.fillStyle = SHEET_FRAME;
  ctx.fillRect(
    placed.x - matPad,
    placed.y - matPad,
    placed.w + matPad * 2,
    placed.h + matPad * 2,
  );
  ctx.drawImage(image, placed.x, placed.y, placed.w, placed.h);

  if (!watermarkActive) return;

  const fontFamily = getPreviewFontFamily(frameParams.fontFamily);
  const captionTop = placed.y + placed.h + canvasH * 0.035;

  // Frame number — ROLL / FRAME labelling reminiscent of contact sheets.
  ctx.fillStyle = SHEET_INK;
  ctx.textBaseline = "alphabetic";
  ctx.textAlign = "left";
  const labelSize = Math.max(11, frameParams.fontSize * 1.05);
  ctx.font = `700 ${labelSize}px "Courier New", "Menlo", ${fontFamily}`;
  ctx.fillText("→ FRAME 24A", placed.x, captionTop + labelSize);

  const camera = config.showCamera
    ? normalizeModel(photo.exif.camera.make, photo.exif.camera.model)
    : "";
  const customLines = sanitizeCustomLines(config.customLines);
  const subline = camera || customLines[0] || "";
  if (subline) {
    ctx.fillStyle = SHEET_MUTED;
    ctx.font = `400 ${Math.max(10, frameParams.fontSize)}px ${fontFamily}`;
    ctx.fillText(subline, placed.x, captionTop + labelSize + Math.max(16, frameParams.fontSize * 1.4));
  }

  // Right: params + date in mono.
  const rightX = placed.x + placed.w;
  const params = config.showParams ? paramsLine(photo.exif, "  ") : "";
  const date = config.showDate ? formatTakenAt(photo.exif.takenAt, config.dateFormat) : "";
  ctx.textAlign = "right";
  if (params) {
    ctx.fillStyle = SHEET_INK;
    ctx.font = `700 ${labelSize}px "Courier New", "Menlo", ${fontFamily}`;
    ctx.fillText(params, rightX, captionTop + labelSize);
  }
  if (date) {
    ctx.fillStyle = SHEET_MUTED;
    ctx.font = `400 ${Math.max(10, frameParams.fontSize)}px "Courier New", "Menlo", ${fontFamily}`;
    ctx.fillText(date, rightX, captionTop + labelSize + Math.max(16, frameParams.fontSize * 1.4));
  }
}

function drawSprocketRow(
  ctx: CanvasRenderingContext2D,
  x: number,
  y: number,
  w: number,
  h: number,
) {
  const holeH = h * 0.55;
  const holeW = h * 0.85;
  const gap = holeW * 1.1;
  const total = holeW + gap;
  const count = Math.floor(w / total);
  const offset = (w - count * total + gap) / 2;
  const holeY = y + (h - holeH) / 2;
  ctx.fillStyle = "#f5f5f5";
  for (let i = 0; i < count; i++) {
    const hx = x + offset + i * total;
    roundedRect(ctx, hx, holeY, holeW, holeH, 2);
    ctx.fill();
  }
}

function roundedRect(
  ctx: CanvasRenderingContext2D,
  x: number,
  y: number,
  w: number,
  h: number,
  r: number,
) {
  ctx.beginPath();
  ctx.moveTo(x + r, y);
  ctx.lineTo(x + w - r, y);
  ctx.quadraticCurveTo(x + w, y, x + w, y + r);
  ctx.lineTo(x + w, y + h - r);
  ctx.quadraticCurveTo(x + w, y + h, x + w - r, y + h);
  ctx.lineTo(x + r, y + h);
  ctx.quadraticCurveTo(x, y + h, x, y + h - r);
  ctx.lineTo(x, y + r);
  ctx.quadraticCurveTo(x, y, x + r, y);
  ctx.closePath();
}

File: src-tauri/src/render/templates/contact_sheet.rs
use image::{imageops::overlay, DynamicImage, Rgba, RgbaImage};

use crate::render::classic_bottom::{
    build_params_line, compute_canvas_size, fill_rect, fill_rounded_rect_solid, fit_photo_in_area,
    format_camera, format_date, resize_photo, ExportExif, ExportFrameParams, ExportTemplateConfig,
};
use crate::render::text::TextRenderer;

const SHEET_BG: Rgba<u8> = Rgba([10, 10, 10, 255]);
const SHEET_INK: Rgba<u8> = Rgba([245, 245, 245, 255]);
const SHEET_MUTED: Rgba<u8> = Rgba([168, 168, 168, 255]);
const SHEET_FRAME: Rgba<u8> = Rgba([255, 255, 255, 255]);
const PERF_COLOR: Rgba<u8> = Rgba([245, 245, 245, 255]);

pub(crate) fn compose_contact_sheet(
    source: DynamicImage,
    frame: &ExportFrameParams,
    exif: &ExportExif,
    config: &ExportTemplateConfig,
    text: Option<&TextRenderer>,
) -> RgbaImage {
    let src_w = source.width();
    let src_h = source.height();
    let (canvas_w, canvas_h) = compute_canvas_size(src_w, src_h, frame);
    let rs = src_w as f32 / 900.0;

    let perf_h = ((canvas_h as f32 * 0.045).round() as u32).max((18.0 * rs) as u32);
    let side_margin = (canvas_w as f32 * 0.07).round() as u32;
    let caption_h = if config.show_watermark {
        ((canvas_h as f32 * 0.14).round() as u32).max((78.0 * rs) as u32)
    } else {
        0
    };

    let area_x = side_margin;
    let area_y = perf_h + (canvas_h as f32 * 0.025).round() as u32;
    let area_w = canvas_w.saturating_sub(side_margin * 2);
    let area_h = canvas_h.saturating_sub(perf_h * 2 + caption_h + (canvas_h as f32 * 0.06).round() as u32);
    let (pw, ph, ox, oy) = fit_photo_in_area(area_w, area_h, src_w, src_h);
    let placed_x = area_x + ox;
    let placed_y = area_y + oy;

    let mut canvas = RgbaImage::from_pixel(canvas_w, canvas_h, SHEET_BG);

    draw_perf_row(&mut canvas, 0, canvas_w, perf_h, rs);
    draw_perf_row(&mut canvas, canvas_h.saturating_sub(perf_h), canvas_w, perf_h, rs);

    let mat_pad = ((canvas_w as f32 * 0.005).round() as u32).max((3.0 * rs) as u32);
    fill_rect(
        &mut canvas,
        placed_x as i64 - mat_pad as i64,
        placed_y as i64 - mat_pad as i64,
        pw + mat_pad * 2,
        ph + mat_pad * 2,
        SHEET_FRAME,
    );

    let photo = resize_photo(source, pw, ph);
    overlay(&mut canvas, &photo, placed_x as i64, placed_y as i64);

    if !config.show_watermark {
        return canvas;
    }

    let Some(rend) = text else { return canvas; };

    let caption_top = placed_y as f32 + ph as f32 + (canvas_h as f32 * 0.035).round();
    let label_pt = (frame.font_size as f32 * 1.05).max(11.0 * rs);
    let label_y = caption_top;

    rend.draw(&mut canvas, "\u{2192} FRAME 24A", placed_x as f32, label_y, label_pt, true, SHEET_INK);

    let camera = if config.show_camera { format_camera(exif) } else { String::new() };
    let subline = if !camera.is_empty() {
        camera
    } else {
        config.custom_lines.first().cloned().unwrap_or_default().trim().to_string()
    };
    if !subline.is_empty() {
        let sub_pt = (frame.font_size as f32).max(10.0 * rs);
        let sy = caption_top + (label_pt * 1.4).max(16.0 * rs);
        rend.draw(&mut canvas, &subline, placed_x as f32, sy, sub_pt, false, SHEET_MUTED);
    }

    let right_x = placed_x as f32 + pw as f32;
    let params = if config.show_params { build_params_line(exif, "  ") } else { String::new() };
    if !params.is_empty() {
        let pw_meas = rend.measure(&params, label_pt, true);
        rend.draw(&mut canvas, &params, right_x - pw_meas, label_y, label_pt, true, SHEET_INK);
    }
    if config.show_date {
        let date = format_date(&exif.taken_at, &config.date_format);
        if !date.is_empty() {
            let sub_pt = (frame.font_size as f32).max(10.0 * rs);
            let dw = rend.measure(&date, sub_pt, false);
            let dy = caption_top + (label_pt * 1.4).max(16.0 * rs);
            rend.draw(&mut canvas, &date, right_x - dw, dy, sub_pt, false, SHEET_MUTED);
        }
    }

    canvas
}

fn draw_perf_row(canvas: &mut RgbaImage, row_y: u32, canvas_w: u32, perf_h: u32, rs: f32) {
    let hole_h = (perf_h as f32 * 0.55).round() as u32;
    let hole_w = (perf_h as f32 * 0.85).round() as u32;
    let gap = (hole_w as f32 * 1.1).round() as u32;
    let total = hole_w + gap;
    if total == 0 { return; }
    let count = canvas_w / total;
    let offset = (canvas_w.saturating_sub(count * total) + gap) / 2;
    let hole_y = row_y + perf_h.saturating_sub(hole_h) / 2;
    let radius = ((hole_w.min(hole_h) as f32 * 0.2).round() as u32).max(1);
    let _ = rs;
    for i in 0..count {
        let hx = offset + i * total;
        fill_rounded_rect_solid(canvas, hx as i64, hole_y as i64, hole_w, hole_h, radius, PERF_COLOR);
    }
}
